src/main/java/com/spikes2212/util/smartmotorcontrollers/SparkFlexWrapper.java

```diff
@@ -0,0 +1,114 @@
+
+import com.revrobotics.spark.ClosedLoopSlot;
+import com.revrobotics.spark.SparkBase;
+import com.revrobotics.spark.SparkFlex;
+import com.revrobotics.spark.SparkFlex;
+import com.revrobotics.spark.config.ClosedLoopConfig;
+import com.revrobotics.spark.config.MAXMotionConfig;
+import com.revrobotics.spark.config.SparkBaseConfig;
+import com.revrobotics.spark.config.SparkFlexConfig;
+import com.spikes2212.control.FeedForwardController;
+import com.spikes2212.control.FeedForwardSettings;
+import com.spikes2212.control.PIDSettings;
+import com.spikes2212.control.TrapezoidProfileSettings;
+import com.spikes2212.util.UnifiedControlMode;
+
+public class SparkFlexWrapper extends SparkFlex implements SmartMotorController {
+
+    private final FeedForwardController feedForwardController;
+    private SparkFlexConfig sparkConfig;
+    private ClosedLoopConfig closedLoopConfig;
+
+    //@TODO change ffcontroller param to relevant enum constant
+    public SparkFlexWrapper(int deviceID, MotorType type, FeedForwardController feedForwardController) {
+        super(deviceID, type);
+        this.feedForwardController = feedForwardController;
+        sparkConfig = new SparkFlexConfig();
+        closedLoopConfig = new ClosedLoopConfig();
+    }
+
+    public SparkFlexWrapper(int deviceID, MotorType type) {
+        this(deviceID, type, new FeedForwardController(FeedForwardSettings.EMPTY_FFSETTINGS,
+                FeedForwardController.DEFAULT_PERIOD));
+    }
+
+    public SparkFlexConfig getSparkConfiguration() {
+        return sparkConfig;
+    }
+
+    public void applyConfiguration(SparkFlexConfig newConfig) {
+        sparkConfig.apply(newConfig);
+        configure(sparkConfig, ResetMode.kNoResetSafeParameters, PersistMode.kNoPersistParameters);
+    }
+
+    public ClosedLoopConfig getClosedLoopConfiguration() {
+        return closedLoopConfig;
+    }
+
+    public void applyClosedLoopConfig(ClosedLoopConfig newConfig) {
+        closedLoopConfig.apply(newConfig);
+        configure(sparkConfig.apply(closedLoopConfig), ResetMode.kNoResetSafeParameters,
+                PersistMode.kNoPersistParameters);
+    }
+
+    public void restoreFactoryDefaults() {
+        sparkConfig = new SparkFlexConfig();
+        closedLoopConfig = new ClosedLoopConfig();
+        configure(sparkConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
+    }
+
+    public void setInverted(boolean inverted) {
+        int leaderID = configAccessor.getFollowerModeLeaderId();
+        if (leaderID != 0) {
+            sparkConfig.follow(leaderID, inverted);
+        } else sparkConfig.inverted(inverted);
+        configure(sparkConfig, ResetMode.kNoResetSafeParameters, PersistMode.kNoPersistParameters);
+    }
+
+    public void follow(SparkBase master) {
+        configure(sparkConfig.follow(master, configAccessor.getInverted()), ResetMode.kNoResetSafeParameters,
+                PersistMode.kNoPersistParameters);
+    }
+
+    public void unfollow() {
```

Comment:
this is fine if you want to keep it, but at what point will you actually use it?

Reply:
rev has it so i said why not because their's angered me